### PR TITLE
Reusing logical query sub-plans (Duplicate Elimination Rule)

### PR DESCRIPTION
**Pull request by mweisgut**
This PR integrates a rule which detects duplicates of logical query sub-plans and eliminates sub-plan duplicates by reusing the first version.
A detailed description follows when this branch in not in WIP modus anymore.

### CODE REVIEW

File: src/lib/optimizer/strategy/duplicate_elimination_rule.cpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+  _adapt_expressions_traversal(node);
+}
+
+void DuplicateEliminationRule::_find_sub_plan_duplicates_traversal(const std::shared_ptr<AbstractLQPNode>& node) const {
+  const auto left_node = node->left_input();
+  const auto right_node = node->right_input();
+
+  if (left_node) {
+    _find_sub_plan_duplicates_traversal(left_node);
```
Comment (mrzzzrm) [inline src/lib/optimizer/strategy/duplicate_elimination_rule.cpp]: Can visit_lqp be used here?

File: src/lib/optimizer/strategy/duplicate_elimination_rule.cpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  }
+  if (right_node) {
+    _find_sub_plan_duplicates_traversal(right_node);
+  }
+  const auto duplicate_iter = std::find_if(_sub_plans.cbegin(), _sub_plans.cend(),
```
Comment (mrzzzrm) [inline src/lib/optimizer/strategy/duplicate_elimination_rule.cpp]: O(n^2) runtime - why not introduce AbstraceLQPNode::hash() or operator<() to be able to use a more efficient data structure?

File: src/lib/optimizer/strategy/duplicate_elimination_rule.cpp (hunk 1 above)
Comment (mrzzzrm) [reply src/lib/optimizer/strategy/duplicate_elimination_rule.cpp]: (less code, no redundant traversal below diamond shapes)

File: src/lib/optimizer/strategy/duplicate_elimination_rule.hpp [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+class AbstractLQPNode;
+
+/**
+ * This optimizer rule eliminates duplicates of the LQP by using sub-plan memoization:
+ * This process uses depth-first traversal and starts integrating sub-plans into the
+ * memoization structure at the leaves of the overall LQP and works upwards to the root.
+ */
+
+class DuplicateEliminationRule : public AbstractRule {
```
Comment (mrzzzrm) [inline src/lib/optimizer/strategy/duplicate_elimination_rule.hpp]: Naming: "DuplicateEliminationRule" immediately made me think of a rule that eliminates duplicate Projections. SubqueryRedundancyEliminationRule or something?

File: src/lib/optimizer/strategy/duplicate_elimination_rule.hpp [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ public:
+  std::string name() const override;
+  void apply_to(const std::shared_ptr<AbstractLQPNode>& node) const override;
+
+ protected:
+  void _find_sub_plan_duplicates_traversal(const std::shared_ptr<AbstractLQPNode>& node) const;
+  void _adapt_expressions_traversal(const std::shared_ptr<AbstractLQPNode>& node) const;
+  mutable std::vector<std::shared_ptr<AbstractLQPNode>> _remaining_stored_table_nodes;
```
Comment (mrzzzrm) [inline src/lib/optimizer/strategy/duplicate_elimination_rule.hpp]: Optimizer might run on multiple plans concurrently, don't store something here, put it into the functions themselves

File: src/lib/expression/expression_utils.hpp [hunk 5]
```diff
@@ -56,6 +56,16 @@ std::vector<std::shared_ptr<AbstractExpression>> expressions_copy_and_adapt_to_d
 std::shared_ptr<AbstractExpression> expression_copy_and_adapt_to_different_lqp(const AbstractExpression& expression,
                                                                                const LQPNodeMapping& node_mapping);
 
+void expressions_adapt_to_different_lqp(std::vector<std::shared_ptr<AbstractExpression>>& expressions,
+                                        const std::vector<std::shared_ptr<AbstractLQPNode>> stored_table_nodes);
```
Comment (mrzzzrm) [inline src/lib/expression/expression_utils.hpp]: constref

File: src/lib/expression/expression_utils.hpp [hunk 6]
```diff
@@ -56,6 +56,16 @@ std::vector<std::shared_ptr<AbstractExpression>> expressions_copy_and_adapt_to_d
...
+
+void expression_adapt_to_different_lqp(std::shared_ptr<AbstractExpression>& expression,
+                                       const std::vector<std::shared_ptr<AbstractLQPNode>> stored_table_nodes);
```
Comment (mrzzzrm) [inline src/lib/expression/expression_utils.hpp]: constref

File: src/lib/expression/expression_utils.hpp [hunk 7]
```diff
@@ -56,6 +56,16 @@ std::vector<std::shared_ptr<AbstractExpression>> expressions_copy_and_adapt_to_d
...
+
+std::shared_ptr<LQPColumnExpression> expression_adapt_to_different_lqp(
+    const LQPColumnExpression& lqp_column_expression,
+    const std::vector<std::shared_ptr<AbstractLQPNode>> stored_table_nodes);
```
Comment (mrzzzrm) [inline src/lib/expression/expression_utils.hpp]: constref

**Comment by mweisgut** [comment]
some issues have to be fixed first...

**Comment by mweisgut** [comment]
Discarded this approach for now because of [this issues](https://github.com/hyrise/hyrise/issues/946#issuecomment-410524521).

**Pull request closed without merging**